### ISSUE TITLE
Cannot get response from /sync/warp protocol

### ISSUE THREAD

**Issue by vikinatora** [paritytech/polkadot#7102]
We are trying to implement the `/sync/warp` protocol in Java. As per [4.8.3. Requesting States](https://spec.polkadot.network/#sect-msg-state-request) we must send 32 bytes that represent the hash of the block encoded with SCALE that we want to start the synchronization from. We are then expecting to receive a response containing some data. However no response is sent back. 

We added some custom debug logs in `warp_request_handler.rs` yet none of them were logged even those at the start of the request handler leading us to believe that it doesn’t get triggered.

Logs:
https://logpaste.com/b2KfwY33

We’ve successfully implemented `/sync/2` and `/light/2` protocols that use protobuf messages meaning that negotiating and communicating over the protocols isn’t an issue.

The command we use to start a local polkadot node:
`./target/debug/polkadot --dev -lsync,sub-libp2p=trace,sync=trance --alice`

The SCALE encoded message we’re sending for hash `0xb71e3ddbfe2b3d1cb534563493b779acbb08ca28019f75cc03c8eeaf55751042` is `b71e3ddbfe2b3d1cb534563493b779acbb08ca28019f75cc03c8eeaf55751042`. After that we send it over the network as a byte array.

Has anyone experienced such an issue?

**Comment by vikinatora** [comment]
Hey @altonen, 

> It looks like it's trying to handle warp sync requests before syncing has started. Are you certain you've enabled appropriate logging targets so it's an actual issue with warp sync?

`Handling warp sync request! 0` is added when the warp request handler is initialized in order to be sure that it should function (the text is misleading, sorry). Here's the handler code:
```
	pub async fn run(mut self) {
		debug!(target: "sync", "Handling warp sync request! 0");
		while let Some(request) = self.request_receiver.next().await {
			debug!(target: "sync", "Handling warp sync request! 1");
			let IncomingRequest { peer, payload, pending_response } = request;

			match self.handle_request(payload, pending_response) {
				Ok(()) => {
					debug!(target: "sync", "Handled grandpa warp sync request from {}.", peer)
				},
				Err(e) => debug!(
					target: "sync",
					"Failed to handle grandpa warp sync request from {}: {}",
					peer, e,
				),
			}
		}
	}

```


> Furthermore, is the request-response substream closed or does it stay open and there's just no bytes to read from it?

The substream stays open indefinately but we close it after 5 seconds, as we expect to receive a response in that time period


> It also looks like it fails to open /block-announces/1 which shouldn't prevent it from responding to warp sync requests though.

`/block-announces/1` isn't implemented yet. It's the next protocol we'll be implementing

**Comment by altonen** [comment]
So you're using the correct protocol name and `multistream-select` succeeds for the substream, then you send the request but it's not received by the warp sync request handler?

If you run the node with `-lsync,sub-libp2p,libp2p=trace`, are you able to see anything interesting in the logs?

**Comment by vikinatora** [comment]
Should the payload be LEB128 prefixed with it's length(32)?

**Comment by altonen** [comment]
What do you mean exactly by the payload? Specification says that all requests and responses must be prefixed with LEB128-encoded length but I don't think warp sync request handler is expecting that so it's dealt with on a lower protocol level

**Comment by vikinatora** [comment]
I'm just asking desperation questions at the moment because we've been stuck on this for the last week and have no idea what we're doing wrong. Our team tried both prepending the message length and not prepending anything. In both cases, we don't get any response.

**Comment by altonen** [comment]
Can you add some prints to the functions of this trait to see if the request is rejected in the codec code: https://github.com/paritytech/substrate/blob/master/client/network/src/request_responses.rs#L926

**Comment by altonen** [comment]
From the logs, assuming this is your warp sync request:
```
2023-04-20 10:54:37.460 TRACE tokio-runtime-worker libp2p_noise::io::framed: read state: Ready
2023-04-20 10:54:37.460 TRACE tokio-runtime-worker libp2p_noise::io::framed: read state: ReadLen { buf: [0, 0], off: 0 }
2023-04-20 10:54:37.507 TRACE tokio-runtime-worker libp2p_noise::io::framed: read state: ReadLen { buf: [0, 0], off: 0 }
2023-04-20 10:54:37.508 TRACE tokio-runtime-worker libp2p_noise::io::framed: read: frame len = 32
2023-04-20 10:54:37.508 TRACE tokio-runtime-worker libp2p_noise::io::framed: read state: ReadData { len: 32, off: 0 }
2023-04-20 10:54:37.508 TRACE tokio-runtime-worker libp2p_noise::io::framed: read: 32/32 bytes
2023-04-20 10:54:37.508 TRACE tokio-runtime-worker libp2p_noise::io::framed: read: decrypting 32 bytes
2023-04-20 10:54:37.508 TRACE tokio-runtime-worker libp2p_noise::io::framed: read: payload len = 0 bytes
```
It says payload length is zero so I think you're crafting the request incorrectly. Warp sync request handler assumes the that request it receives is 32 bytes so shouldn't the encrypted payload which also contains the payload length then have to be larger than 32 bytes.

If this is the case, in your position I'd look at the wire format of a warp sync request sent by Substrate and compare that to the request you're sending. Maybe you're missing the length of the request or you're encoding it incorrectly.

**Comment by vikinatora** [comment]
Yes, that was the problem. We were encoding the request correctly, but incorrectly for transferring over the network (netty specific issue). 
Thank you for the guidance, it was invaluable for us @altonen !
